### DISCUSSION TITLE
createStorage not working with Next's static exports

### DISCUSSION THREAD

**Discussion by quangnmwork** [pmndrs/jotai#2620] [Bug report]
## Summary

```
ReferenceError: sessionStorage is not defined

 const storage = createJSONStorage<string>(() => sessionStorage);

```

When using jotai's `createStorage` function with static export of [Nextjs
](https://nextjs.org/docs/pages/building-your-application/deploying/static-exports)

I got the above error. I don't know where I went wrong. Is this a bug of jotai?

## Link to reproduction

https://github.com/quangnmwork/nextjs-bug-err-with-jotai

```
Next: 14.2.4
jotai: 2.8.3
```

In `next.config.mjs`

```
const nextConfig = {
  reactStrictMode: true,
  output: "export",
};

```

**Comment by dai-shi** [comment]
Is it the same with localStorage? How about window.sessionStorage?

I wonder if anyone can help here.

**Reply by quangnmwork** [reply]
The same error with localStorage

**Reply by dai-shi** [reply]
Thanks. Can you try some older versions of Jotai and see if they change anything?

**Reply by quangnmwork** [reply]
Hi, it still not work with other version.

**Reply by quangnmwork** [reply]
Currently i'm still doing some workaround for this problem . 

The thing I doing is write  some custom storage for `atomWithStorage` : 

```
getItem(key, initialValue) {
      // try to not get error here
      if (typeof window === 'undefined') {
        return initialValue;
      }

      const storedValue = sessionStorage.getItem(key);
      try {
        return schema.parse(JSON.parse(storedValue ?? ''));
      } catch {
        return initialValue;
      }
    },
```

**Reply by dai-shi** [reply]
Thanks for investigating.

Hmm, interesting. Maybe it has never worked before.
I wonder if anyone is interested in investing further. Help wanted.

Meanwhile, your workaround with custom storage seems good.

**Comment by mishrasatyam** [comment]
Hi @quangnmwork 

It seems like a next js issue where ssr is enabled in `output: "export"` mode.

Another workound is calling moving `useAtomValue` to a seperate component and calling it dynamically in `index`

`HomeClientComponent.tsx` in `src/components/HomeClientComponent.tsx`
```js
import { dummyAtom } from "@/pages/jotai";
import { useAtomValue } from "jotai";

export default function HomeClientComponent() {
  const dummyValue = useAtomValue(dummyAtom);
  return <>Value: {dummyValue}</>;
}
```

`index.tsx`
```js
import dynamic from "next/dynamic";

const DynamicHome = dynamic(() => import("../components/HomeCLientComponent"), {
  loading: () => <p>Loading...</p>,
  ssr: false,
});

export default function Home() {
  return <DynamicHome />;
}

```

**Comment by himself65** [comment]
in static export mode, the component use storageAtom shouldn't be generated because it requires web API in the runtime. So I think the good fix is to Suspense the component

**Comment by wojtekmaj** [comment]
I have a Next.js-compatible helper like so:

```ts
import { atom } from "jotai";
import { atomWithStorage, createJSONStorage } from "jotai/utils";

const isSSR = typeof window === "undefined";

/**
 * Creates a Jotai atom that uses session storage for persistence in the browser.
 * Falls back to a regular atom with an initial value when executed in a server-side environment.
 *
 * @template T The type of the value stored in the atom.
 * @param key The key used to store the value in session storage.
 * @param initialValue The initial value of the atom.
 * @returns A Jotai atom that persists its value in session storage or behaves as a regular atom in SSR.
 */
export const ssrAtomWithStorage = <T>(key: string, initialValue: T) => {
  return isSSR
    ? atom<T>(initialValue)
    : atomWithStorage<T>(
        key,
        initialValue,
        createJSONStorage(() => sessionStorage),
      );
};
```
